plugins/cef/src/plugin.cpp

```diff
@@ -121,6 +105,17 @@ class reader : public format::single_layout_reader {
       auto msg = to<message>(std::string_view{line});
       if (!msg)
         return msg.error();
+      // TODO: Resetting the builder for every line is highly inefficient. For
+      // heterogeneous CEF, we have no other way. We could add a flag that skips
+      // the check after the first record. Or we could skip the check if the
+      // user provided a custom schema.
```

Comment:
You could not use the table slice builder interface but instead create the individual arrays and create a union of all inferred types. When you encounter a new field for the first time, add nulls for all previous rows for that field. If a row doesn't have a field that a previous row had, add a null for it as well.

If you want to do type inference here rather than letting the user provide a schema this'd be the way to go.

Ultimately I don't think it'd be right to implement this locally within this reader, as that will be hard to maintain. It should be a custom table slice builder instead that allows for creating the schema dynamically like this so it's orthogonal to the reader API.

Reply:
> Ultimately I don't think it'd be right to implement this locally within this reader, as that will be hard to maintain. It should be a custom table slice builder instead that allows for creating the schema dynamically like this so it's orthogonal to the reader API.

This. There needs to be an operator for data with high schema variance, similar to Zed's [fuse](https://zed.brimdata.io/docs/v1.3.0/language/operators/fuse).

I don't expect this to be a high-volume format for now, so I think it's fine to ship this first as is, and then address these orthogonal schema inference concerns independently.

Reply:
I don't think this is a separate operator, but rather a separate builder and then an option for every operator that utilizes a builder. It cannot feasibly be its own operator as an operator has a single input and a single output type, and I don't think we can define the input type well here.

We have the tools in the code base already to write such a new builder. I could imagine it having some API like this:

```c++
class schemaless_builder {
  // Create a new schemaless builder; requires setting the schema name.
  explicit schemaless_builder(std::string name) { ... }

  // Add an entry to the current row. Requires explicitly setting the name of the field
  // to add. Modifies the schema if the field does not already exist. A dot is
  // interpreted as a nested record.
  template <basic_type Type>
  auto add(std::string_view field, const Type& type, const view<Type>& data) { ... }

  // TODO: some custom interface for map and list

  // Move on to the next row.
  auto finish_row() { ... }

  // Create a table slice.
  auto finish() -> table_slice { ... }
};
```

Reply:
I think this is the wrong place for continuing discussing a solution, as results get buried here. Maybe the community Slack?